bin/node/src/pruning_config.rs

```diff
@@ -0,0 +1,151 @@
+    fn pruning_changed(params: &PruningParams) -> bool {
+        let state_pruning_changed =
+            params.state_pruning.unwrap_or(DEFAULT_STATE_PRUNING) != DEFAULT_STATE_PRUNING;
+        let blocks_pruning_changed = params.blocks_pruning != DEFAULT_BLOCKS_PRUNING;
+
+        state_pruning_changed || blocks_pruning_changed
+    }
+
+    pub fn validate_and_fix_pruning_options(cli: &mut Cli) -> Self {
+        let overwritten_pruning = Self::pruning_changed(&cli.run.import_params.pruning_params);
+        let pruning_enabled = cli.aleph.enable_pruning();
+
+        let mut result = PruningConfigValidator {
+            pruning_enabled,
+            overwritten_pruning,
+            invalid_state_pruning_setting: Ok(()),
+            invalid_blocks_pruning_setting: Ok(()),
+            invalid_database_backend: Ok(()),
+        };
+
+        if !pruning_enabled {
+            // We need to override state pruning to our default (archive), as substrate has 256 by default.
+            // 256 does not work with our code.
+            cli.run.import_params.pruning_params.state_pruning = Some(DEFAULT_STATE_PRUNING);
+            cli.run.import_params.pruning_params.blocks_pruning = DEFAULT_BLOCKS_PRUNING;
+            return result;
+        }
+
+        match cli
+            .run
+            .import_params
+            .pruning_params
+            .state_pruning
+            .get_or_insert(DatabasePruningMode::Custom(MINIMAL_STATE_PRUNING))
+        {
+            DatabasePruningMode::Archive | DatabasePruningMode::ArchiveCanonical => {}
+            DatabasePruningMode::Custom(max_blocks) => {
+                if *max_blocks < MINIMAL_STATE_PRUNING {
+                    result.invalid_state_pruning_setting = Err(*max_blocks);
+                    *max_blocks = MINIMAL_STATE_PRUNING;
+                }
+            }
+        }
+
+        match cli.run.import_params.pruning_params.blocks_pruning {
+            DatabasePruningMode::Archive | DatabasePruningMode::ArchiveCanonical => {}
+            DatabasePruningMode::Custom(blocks_pruning) => {
+                result.invalid_blocks_pruning_setting = Err(blocks_pruning);
+                cli.run.import_params.pruning_params.blocks_pruning = DEFAULT_BLOCKS_PRUNING;
+            }
+        }
+
+        match cli
+            .run
+            .import_params
+            .database_params
+            .database
+            .get_or_insert(DEFAULT_DATABASE_FOR_PRUNING)
+        {
+            Database::ParityDb => {}
+            db @ (Database::RocksDb | Database::Auto | Database::ParityDbDeprecated) => {
+                result.invalid_database_backend = Err(());
+                *db = DEFAULT_DATABASE_FOR_PRUNING;
+            }
+        }
+        result
+    }
+
+    pub fn report_pruning_validation_result(self) {
+        if !self.pruning_enabled {
+            if self.overwritten_pruning {
+                warn!("Pruning not enabled. Switching to keeping all block bodies and states.");
```

Comment:
Maybe add instructions how to actually enable it, i.e. "Please use --enable-pruning to...".

Reply:
And maybe panic in such case

Reply:
I'm satisfied with just the change you made now, but if you want to panic, panicking on every misconfiguration separately is also fine, there usually won't be that many of them – and this is what happens if you misspell multiple CLI arguments anyway, don't ask me how I know. ;p

Reply:
ok let's leave it as it is